### ISSUE TITLE
Writing large object to file is slow

### ISSUE THREAD

**Issue by borkdude** [cognitect/transit-clj#43]
Writing a large object to a file with transit-clj was slower than I expected. It takes about 1 second with transit-clj and roughly 1/10th of that time with regular `spit`.

Test data:
[test.edn.zip](https://github.com/cognitect/transit-clj/files/3024181/test.edn.zip)
Extract to test.edn.

Repro:
```
clj -Sdeps '{:deps {com.cognitect/transit-clj {:mvn/version "0.8.313"}}}'
(require '[clojure.edn :as edn])
(def edn (edn/read-string (slurp "test.edn")))
(count (keys edn)) ;;=> 799
(require '[cognitect.transit :as transit])
(require '[clojure.java.io :as io])
(def writer (transit/writer (io/output-stream (io/file "transit.json")) :json))
(time (transit/write writer edn)) ;; prints:
"Elapsed time: 1151.116438 msecs"
```

While writing the same EDN just with `spit` is much faster:
```
(time (with-open [fos (java.io.FileOutputStream. "/tmp/foo.edn")] (let [w (io/writer fos)] (.write w (str edn)) (.flush fos)))) ;; prints:
"Elapsed time: 73.316135 msecs"
``` 

Flamegraph of transit:
[transit-flamegraph.svg.zip](https://github.com/cognitect/transit-clj/files/3024193/transit-flamegraph.svg.zip)

Flamegraph of EDN:
[edn-flamegraph.svg.zip](https://github.com/cognitect/transit-clj/files/3024194/edn-flamegraph.svg.zip)

**Comment by strssndktn** [comment]
@borkdude I investigated this a bit and it seems that transit calls `.flush` overly aggressive, thus forcing the data to be flushed out to disk on every element (I am not completely sure how and when `.flush` gets called or what are the reasons for that). As an experiment, could you put a proxy in between your output-stream and transit/writer overwriting `.flush` with a noop function?

**Comment by borkdude** [comment]
@strssndktn I worked around this by first writing to a `ByteArrayOutputStream` for which `.flush` is a no-op and that helped a lot.

**Comment by tonsky** [comment]
I’m hitting the same issue. This did the trick for me:

```
(defn no-flush-output-stream [^OutputStream os]
  (proxy [BufferedOutputStream] [os]
    (flush [])
    (close []
      (.flush os)
      (.close os))))
```

**Comment by tonsky** [comment]
UPD: that actually should’ve been

```
(defn no-flush-output-stream [^OutputStream os]
  (proxy [BufferedOutputStream] [os]
    (flush [])
    (close []
      (proxy-super flush)
      (proxy-super close))))
```

**Comment by ericdallo** [comment]
@tonsky I just updated clojure-lsp to use your suggestion and improved memory usage indeed for huge maps, thanks for that!

**Comment by borkdude** [comment]
Thanks. I needed to add a couple more type hints to make it work without reflection:

``` clojure
(defn no-flush-output-stream
  "See https://github.com/cognitect/transit-clj/issues/43#issuecomment-1650341353"
  ^java.io.OutputStream [^java.io.OutputStream os]
  (proxy [java.io.BufferedOutputStream] [os]
    (flush [])
    (close []
      (let [^java.io.BufferedOutputStream this this]
        (proxy-super flush)
        (proxy-super close)))))
```

**Comment by tonsky** [comment]
Did the exact same change yesterday myself! Look at that, two senior Clojure devs got 10-line function right after just 3 attempts!

**Comment by tonsky** [comment]
Well, the problem is not that transit flushes at the end of writing an entity. Problem is that it flushes multiple times in the middle of writing entity. I don’t think that could serve any purpose, it must’ve been an implementation bug

**Comment by timewald** [comment]
It flushes for any complex entity it writes, including those that are
nested. So if you have an array of 100 maps, it will flush 101 times - once
for each map and once for the array. From a network streaming perspective,
I think that's preferable, but I can see the other perspective for sure.

**Comment by tonsky** [comment]
Why is it preferable from network perspective? All these maps are wrapped in a list, and you can’t read part of the list on the other side, can you?

**Comment by timewald** [comment]
You can start the process of reading on the other side. The Jackson parser
is stream oriented, so while the transit reader won't produce a result
until, say, an entire array has arrived, it can be reading and decoding the
bytes while the writer is still producing output.

**Comment by tonsky** [comment]
Sure, but that’s why you wrap your OutputStream in BufferedOutputStream. It’ll buffer enough for you and flush itself when buffer runs out

**Comment by timewald** [comment]
Yeah we could have done that too. I can't honestly remember at this point
if we considered it or not. :)
